MsWheaPkg/Crates/RustMuTelemetryHelperLib/src/lib.rs

```diff
@@ -0,0 +1,229 @@
+pub fn log_telemetry(
+    is_fatal: bool,
+    class_id: EfiStatusCodeValue,
+    extra_data1: u64,
+    extra_data2: u64,
+    component_id: Option<&efi::Guid>,
+    library_id: Option<&efi::Guid>,
+    ihv_id: Option<&efi::Guid>,
+) -> Result<(), efi::Status> {
+    log_telemetry_internal(
+        &BOOT_SERVICES,
+        is_fatal,
+        class_id,
+        extra_data1,
+        extra_data2,
+        component_id,
+        library_id,
+        ihv_id,
+    )
+}
+
+fn log_telemetry_internal<B: BootServices>(
+    boot_services: &B,
+    is_fatal: bool,
+    class_id: EfiStatusCodeValue,
+    extra_data1: u64,
+    extra_data2: u64,
+    component_id: Option<&efi::Guid>,
+    library_id: Option<&efi::Guid>,
+    ihv_id: Option<&efi::Guid>,
+) -> Result<(), efi::Status> {
+    let status_code_type: EfiStatusCodeType =
+        if is_fatal { MS_WHEA_ERROR_STATUS_TYPE_FATAL } else { MS_WHEA_ERROR_STATUS_TYPE_INFO };
+
+    let error_data = MsWheaRscInternalErrorData {
+        library_id: *library_id.unwrap_or(&guid::ZERO),
+        ihv_sharing_guid: *ihv_id.unwrap_or(&guid::ZERO),
+        additional_info_1: extra_data1,
+        additional_info_2: extra_data2,
+    };
+
+    StatusCodeRuntimeProtocol::report_status_code(
+        boot_services,
+        status_code_type,
+        class_id,
+        0,
+        component_id,
+        MS_WHEA_RSC_DATA_TYPE_GUID,
+        error_data,
+    )
+}
+
+#[cfg(not(tarpaulin_include))]
+pub fn init_telemetry(efi_boot_services: &efi::BootServices) {
+    BOOT_SERVICES.initialize(efi_boot_services)
+}
+
+#[cfg(test)]
+mod test {
+    use boot_services::MockBootServices;
+    use mu_pi::protocols::status_code;
+    use mu_pi::protocols::status_code::{EfiStatusCodeData, EfiStatusCodeType, EfiStatusCodeValue};
+    use mu_rust_helpers::guid::{guid, guid_fmt};
+    use r_efi::efi;
+    use rust_advanced_logger_dxe::{debugln, DEBUG_INFO};
+    use uuid::uuid;
+
+    use crate::{
+        log_telemetry_internal,
+        status_code_runtime::StatusCodeRuntimeProtocol,
+        MsWheaRscInternalErrorData, MS_WHEA_ERROR_STATUS_TYPE_FATAL,
+    };
+    use core::mem::size_of;
+
+    const DATA_SIZE: usize = size_of::<EfiStatusCodeData>() + size_of::<MsWheaRscInternalErrorData>();
+    const MOCK_CALLER_ID: efi::Guid = guid!("d0d1d2d3-d4d5-d6d7-d8d9-dadbdcdddedf");
+    const MOCK_STATUS_CODE_VALUE: EfiStatusCodeValue = 0xa0a1a2a3;
+
+    extern "efiapi" fn mock_report_status_code(
+        r#type: EfiStatusCodeType,
+        value: EfiStatusCodeValue,
+        instance: u32,
+        caller_id: *const efi::Guid,     // Optional
+        _data: *const EfiStatusCodeData, // Optional
+    ) -> efi::Status {
+        assert_eq!(r#type, MS_WHEA_ERROR_STATUS_TYPE_FATAL);
+        assert_eq!(value, MOCK_STATUS_CODE_VALUE);
+        assert_eq!(instance, 0);
+        assert_eq!(unsafe { *caller_id }, MOCK_CALLER_ID);
+        debugln!(DEBUG_INFO, "[MockStatusCodeRuntime] caller_id: {}", guid_fmt!(unsafe { *caller_id }));

```

Comment:
Probably don't need this print in the unit test

Reply:
Removed